### PR TITLE
refactor smoothL1Criterion and fix tensor storage usage

### PR DESCRIPTION
**Pull request by SeaOfOcean**
## What changes were proposed in this pull request?

refactor smoothL1Criterion and fix tensor storage usage

## How was this patch tested?

unit test

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SmoothL1Criterion.scala [hunk 1]
```diff
@@ -41,23 +41,21 @@ import scala.reflect.ClassTag
 class SmoothL1Criterion[@specialized(Float, Double) T: ClassTag](sizeAverage: Boolean = true)
                                     (implicit ev: TensorNumeric[T])
   extends TensorCriterion[T] {
-  @transient var buffer: Tensor[T] = null
+  val buffer: Tensor[T] = Tensor[T]()
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SmoothL1Criterion.scala]: add clear state

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SmoothL1Criterion.scala [hunk 2]
```diff
@@ -41,23 +41,21 @@ import scala.reflect.ClassTag
...
     require(input.nElement() == target.nElement())
-    if (buffer == null) {
-      buffer = Tensor[T]()
-    }
     buffer.resizeAs(input).copy(input)
     buffer.add(ev.fromType(-1), target).abs()
     val data = buffer.storage().array()
-    for (i <- 0 until data.length) {
+    var i = 0
+    while (i < buffer.nElement()) {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SmoothL1Criterion.scala]: need offset too?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SmoothL1Criterion.scala [hunk 3]
```diff
@@ -41,23 +41,21 @@ import scala.reflect.ClassTag
...
       if (ev.isGreater(ev.fromType(1), data(i))) {
         data(i) = ev.times(ev.fromType[Double](0.5), ev.times(data(i), data(i)))
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SmoothL1Criterion.scala]: need offset too?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SmoothL1Criterion.scala [hunk 4]
```diff
@@ -70,22 +68,22 @@ class SmoothL1Criterion[@specialized(Float, Double) T: ClassTag](sizeAverage: Bo
...
-        data(i) = ev.negative(norm)
+    val offset = gradInput.storageOffset() - 1
+    var i = 0
+    while (i < gradInput.nElement()) {
+      if (ev.isGreater(ev.fromType(-1), data(i + offset))) {
+        data(i + offset) = ev.negative(norm)
       }
-      else if (ev.isGreater(data(i), ev.fromType(1))) {
-        data(i) = norm
+      else if (ev.isGreater(data(i + offset), ev.fromType(1))) {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SmoothL1Criterion.scala]: ev.one instead of ev.fromType(1)

**Comment by zhangxiaoli73** [comment]
add some unit test to cover this change

**Comment by SeaOfOcean** [comment]
pr validate pass

**Comment by zhangxiaoli73** [comment]
Code LGTM, perhaps you should add some unit test to cover this change

**Comment by zhangxiaoli73** [comment]
LGTM

**Pull request merged**